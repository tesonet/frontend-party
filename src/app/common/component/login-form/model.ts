export interface ILogin {
	response: any;
	token: string;
	message: {};
	loading: boolean;
}

export interface ILoginMessage {
	type: string;
	text: string;
}
