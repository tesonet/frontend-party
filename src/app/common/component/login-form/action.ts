import { Action } from '@ngrx/store';

export const ActionTypes = {
	LOGIN: '[Login] Login',
	LOGIN_SUCCESS: '[Login] Login success',
	LOGIN_FAILED: '[Login] Login failed',
	LOGIN_CLEAR: '[Login] Login clear',
};

export class LoginAction implements Action {
	type = ActionTypes.LOGIN;
	constructor(public payload: any) { }
}

export class LoginSuccessAction implements Action {
	type = ActionTypes.LOGIN_SUCCESS;
	constructor(public payload: any) { }
}

export class LoginFailedAction implements Action {
	type = ActionTypes.LOGIN_FAILED;
	constructor(public payload: any) { }
}

export class LoginClearAction implements Action {
	type = ActionTypes.LOGIN_CLEAR;
	constructor() { }
}

export type Actions
	= LoginAction
	| LoginSuccessAction
	| LoginFailedAction
	| LoginClearAction
