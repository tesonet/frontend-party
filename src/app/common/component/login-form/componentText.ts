export const LABEL_LIST = {
	'userName': 'Username',
	'password': 'Password',
	'logIn': 'Log In'
};

export const MESSAGE_LIST = {
	'errorMessageUserName': 'User name is required',
	'errorMessagePassword': 'Password is required',
	'loginSuccessful': 'Login successful',
	'loginFailed': 'Wrong username or password'
};
