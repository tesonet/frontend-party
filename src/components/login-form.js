import React, {useState} from "react";
import {connect} from "react-redux";
import styled from 'styled-components';

import LoginInput from "../components/login-input";
import LoginButton from "../components/login-button";
import {setLogged} from "../redux/actions";
import {login} from "../api/login";
import LoginError from "../components/login-error";
import {device_max_width} from "../config/device-breakpoints";


const Form = styled.form`
    display: flex;
    flex-direction: column;
    @media only screen and ${device_max_width.mobileL} {
         width: 100%;
    }
  `;

export const LoginForm = (props) => {
    const [username, setUsername] = useState("");
    const [password, setPassword] = useState("");
    const [error, setError] = useState("");

    const validateForm = () => {
        return username.length > 0 && password.length > 0;
    };

    const handleSubmit = (event) => {
        event.preventDefault();
        if (validateForm()) {
            login({username, password})
                .then(json => {
                    props.setLogged(true);
                    localStorage.setItem("token", json.token);
                })
                .catch(() => {
                    setPassword("");
                    setError("Invalid credentials");
                });
        }
    };

    return (
        <Form onSubmit={handleSubmit}>
            <LoginInput label='Username' value={username} onChange={setUsername}/>
            <LoginInput label='Password' value={password} onChange={setPassword}/>
            <LoginError error={error}/>
            <LoginButton/>
        </Form>
    );
};

export default connect(
    null,
    {setLogged}
)(LoginForm);
