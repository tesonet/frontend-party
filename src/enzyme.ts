import { configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import fetchMock from "jest-fetch-mock";

declare let global: { fetch: {} };

global.fetch = fetchMock;
configure({ adapter: new Adapter() });
