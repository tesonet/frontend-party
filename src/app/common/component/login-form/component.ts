import { Component, ViewChild, OnDestroy } from '@angular/core';
import { Router } from '@angular/router';
import { Store } from '@ngrx/store';
import { Observable } from 'rxjs/Observable';
import { Subscription } from 'rxjs/Subscription';
import { IAppState } from '../../../global';
import * as login from './action';
import { LoginFormTemplate } from './template';
import * as ComponentText from './component-text';
import * as messageType from './model';

@Component({
	selector: 'app-login-form',
	templateUrl: './component.html'
})
export class LoginFormComponent implements OnDestroy {

	private loginSubscription: Subscription;

	public login$: Observable<any>;
	public componentText = ComponentText;
	public model: LoginFormTemplate = new LoginFormTemplate();
	public message: messageType.ILoginMessage;
	@ViewChild('loginForm') public loginForm: any;

	constructor(
		public store: Store<IAppState>,
		private router: Router
	) {
		this.login$ = this.store.select(state => state.login);
		this.loginSubscription = this.login$.subscribe(data => {
			this.message = data.message;
			if (data.message && data.message.type && data.message.type === 'success') {
				this.navigateToServerList();
			}
		});
	}

	public onSubmit() {
		this.removeMessage();
		if (this.loginForm.valid) {
			this.store.dispatch(
				new login.LoginAction(
					{
						userLoginData: {
							'username': this.loginForm.value.credentials.userName,
							'password': this.loginForm.value.credentials.password
						}
					}
				)
			);
			this.loginForm.reset();
		}
	}

	private removeMessage() {
		if (this.message) {
			this.message = null;
		}
	}

	private clearState() {
		this.store.dispatch(
			new login.LoginClearAction()
		);
	}

	private clearLogin() {
		this.removeMessage();
		this.clearState();
		this.loginSubscription.unsubscribe();
	}

	private navigateToServerList() {
		setTimeout(
			() => {
				this.clearLogin();
				this.router.navigate(
					['/server-list'],
					{}
				);
			},
			1000
		)
	}

	ngOnDestroy() {
		this.clearLogin();
	}

}
