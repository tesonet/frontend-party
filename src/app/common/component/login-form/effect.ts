import { Actions, Effect } from '@ngrx/effects';
import { Http, Headers, RequestOptions } from '@angular/http';
import { Injectable } from '@angular/core';
import { of } from 'rxjs/observable/of';
import { CONFIG } from '../../../global';
import '../../rxjs/common-imports';
import { ActionTypes, LoginSuccessAction, LoginFailedAction } from './action';
import { SessionStorageService, TokenSecureService } from '../../../common/service'

@Injectable()
export class LoginEffect {

	@Effect() Login$ =
		this.actions$.ofType(ActionTypes.LOGIN)
			.debounceTime(CONFIG.dev.fakeDelay)
			.map(action => action.payload)
			.switchMap(({userLoginData}) => {
				return this.login(userLoginData)
					.map(
						res => {
							this.setSessionStorageToken(res.json().token);
							return new LoginSuccessAction(res.json());
						}
					)
					.catch(
						(e) => of(new LoginFailedAction(e))
					);
			});

	constructor(
		private http: Http,
		private actions$: Actions,
		private sessionStorageService: SessionStorageService,
		private tokenSecureService: TokenSecureService
	) { }

	private login(userLoginData) {
		const headers = new Headers({ 'Content-Type': 'application/json' });
		const options = new RequestOptions({ headers: headers });
		const route = CONFIG.api.root + CONFIG.api.login;
		return this.http.post(
			route,
			userLoginData,
			options
		);
	}

	private setSessionStorageToken(token) {
		const securedToken = this.tokenSecureService.secure(token);
		this.sessionStorageService.setItem('token', securedToken);
	}

}
