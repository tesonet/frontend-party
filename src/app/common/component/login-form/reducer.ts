import { Action } from '@ngrx/store';
import * as login from './action';
import * as ComponentText from './component-text';
import * as messageType from '../../../global/model/message';

export interface State {
	response: any;
	token: string;
	message: messageType.IMessage;
	loading: boolean;
}

const initialState: State = {
	response: null,
	token: null,
	message: null,
	loading: false
};

function handleLogin(state: State, action: Action): State {
	return Object.assign(
		{},
		state,
		{
			message: null,
			loading: true
		}
	);
}

function handleLoginSuccess(state: State, action: Action): State {
	return Object.assign(
		{},
		state,
		{
			token: action.payload.token,
			message: {
				'type': 'success',
				'text': ComponentText.MESSAGE_LIST.loginSuccessful
			},
			loading: false
		}
	);
}

function handleLoginFailed(state: State, action: Action): State {
	return Object.assign(
		{},
		state,
		{
			response: action.payload,
			message: {
				'type': 'failed',
				'text': ComponentText.MESSAGE_LIST.loginFailed
			},
			loading: false
		}
	)
}
function handleLoginClear(state: State, action: Action): State {
	return Object.assign(
		{},
		state,
		{
			response: null,
			token: null,
			message: null,
			loading: false
		}
	)
}

export function reducer(state = initialState, action: Action): State {
	switch (action.type) {
		case login.ActionTypes.LOGIN:
			return handleLogin(state, action);
		case login.ActionTypes.LOGIN_SUCCESS:
			return handleLoginSuccess(state, action);
		case login.ActionTypes.LOGIN_FAILED:
			return handleLoginFailed(state, action);
		case login.ActionTypes.LOGIN_CLEAR:
			return handleLoginClear(state, action);
		default:
			return state;
	}
}
