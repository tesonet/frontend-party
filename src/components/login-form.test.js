import React from "react";
import {shallow} from "enzyme";
import toJson from "enzyme-to-json";

import {LoginForm} from "./login-form";


const component = shallow(<LoginForm/>);

it('renders correctly', () => {
    expect(toJson(component)).toMatchSnapshot();
});
