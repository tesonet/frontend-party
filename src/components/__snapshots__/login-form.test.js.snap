// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<styled.form
  onSubmit={[Function]}
>
  <LoginInput
    label="Username"
    onChange={[Function]}
    value=""
  />
  <LoginInput
    label="Password"
    onChange={[Function]}
    value=""
  />
  <LoginError
    error=""
  />
  <LoginButton />
</styled.form>
`;
